File: backend/src/services/eventRegistrationService.js
/**
 * Event Registration Service - Firebase Realtime Database
 */

const { v4: uuidv4 } = require('uuid');
const { getAll, getById, create, update, remove, queryByChild, db } = require('../config/firebase');

const REGISTRATIONS_REF = 'eventRegistrations';
const EVENTS_REF = 'events';
const USERS_REF = 'users';

class EventRegistrationService {
    /**
     * Get all registrations (admin)
     */
    async findAll() {
        const registrations = await getAll(REGISTRATIONS_REF);
        // Enrich with event and user data
        return Promise.all(registrations.map(async (reg) => {
            const event = await getById(EVENTS_REF, reg.eventId);
            const user = await getById(USERS_REF, reg.userId);
            return {
                ...reg,
                eventTitle: event?.title || null,
                eventDate: event?.eventDate || null,
                eventLocation: event?.location || null,
                userName: user?.name || null,
                userEmail: user?.email || null,
            };
        }));
    }

    /**
     * Get registrations by event
     */
    async findByEvent(eventId) {
        const registrations = await queryByChild(REGISTRATIONS_REF, 'eventId', eventId);
        return Promise.all(registrations.map(async (reg) => {
            const user = await getById(USERS_REF, reg.userId);
            return {
                ...reg,
                userName: user?.name || null,
                userEmail: user?.email || null,
            };
        }));
    }

    /**
     * Get registrations by user
     */
    async findByUser(userId) {
        const registrations = await queryByChild(REGISTRATIONS_REF, 'userId', userId);
        return Promise.all(registrations.map(async (reg) => {
            const event = await getById(EVENTS_REF, reg.eventId);
            return {
                ...reg,
                eventTitle: event?.title || null,
                eventDate: event?.eventDate || null,
                eventLocation: event?.location || null,
                eventImageUrl: event?.imageUrl || null,
            };
        }));
    }

    /**
     * Check if user is registered for event
     */
    async isRegistered(eventId, userId) {
        const registrations = await queryByChild(REGISTRATIONS_REF, 'eventId', eventId);
        return registrations.some(reg => reg.userId === userId);
    }

    /**
     * Register user for event
     */
    async register(eventId, userId) {
        // Check if already registered
        const isAlreadyRegistered = await this.isRegistered(eventId, userId);
        if (isAlreadyRegistered) {
            throw new Error('Anda sudah terdaftar di event ini');
        }

        const id = uuidv4();
        const now = new Date().toISOString();

        const registrationData = {
            eventId,
            userId,
            status: 'registered', // registered, attended, cancelled
            registeredAt: now,
            updatedAt: now,
        };

        await create(REGISTRATIONS_REF, id, registrationData);
        return { id, ...registrationData };
    }

    /**
     * Cancel registration
     */
    async cancel(id, userId) {
        const registration = await getById(REGISTRATIONS_REF, id);
        if (!registration) {
            throw new Error('Pendaftaran tidak ditemukan');
        }
        if (registration.userId !== userId) {
            throw new Error('Tidak memiliki akses untuk membatalkan pendaftaran ini');
        }

        await update(REGISTRATIONS_REF, id, {
            status: 'cancelled',
            updatedAt: new Date().toISOString()
        });
        return { success: true };
    }

    /**
     * Update registration status (admin)
     */
    async updateStatus(id, status) {
        await update(REGISTRATIONS_REF, id, {
            status,
            updatedAt: new Date().toISOString(),
        });
        return getById(REGISTRATIONS_REF, id);
    }

    /**
     * Delete registration (admin)
     */
    async delete(id) {
        await remove(REGISTRATIONS_REF, id);
        return { deleted: true };
    }

    /**
     * Get registration count for event
     */
    async getCountByEvent(eventId) {
        const registrations = await queryByChild(REGISTRATIONS_REF, 'eventId', eventId);
        return registrations.filter(r => r.status === 'registered').length;
    }
}

module.exports = new EventRegistrationService();

File: frontend/lib/services/event_registration_service.dart
import 'api_client.dart';

class EventRegistrationService {
  EventRegistrationService._();
  static final EventRegistrationService instance = EventRegistrationService._();
  final ApiClient _client = ApiClient.instance;

  /// Register for event
  Future<EventRegistrationModel> register(String eventId) async {
    final response = await _client.post('/event-registrations/$eventId/register', {});
    return EventRegistrationModel.fromJson(response['registration'] as Map<String, dynamic>);
  }

  /// Get user's registrations
  Future<List<EventRegistrationModel>> fetchMyRegistrations() async {
    final response = await _client.get('/event-registrations/my-registrations');
    final registrations = response['registrations'] as List<dynamic>? ?? [];
    return registrations
        .map((json) => EventRegistrationModel.fromJson(json as Map<String, dynamic>))
        .toList();
  }

  /// Check if registered for event
  Future<bool> checkRegistration(String eventId) async {
    final response = await _client.get('/event-registrations/$eventId/check');
    return response['isRegistered'] as bool? ?? false;
  }

  /// Cancel registration
  Future<void> cancelRegistration(String registrationId) async {
    await _client.post('/event-registrations/$registrationId/cancel', {});
  }

  /// Get all registrations (admin)
  Future<List<EventRegistrationModel>> fetchAllRegistrations() async {
    final response = await _client.get('/event-registrations');
    final registrations = response['registrations'] as List<dynamic>? ?? [];
    return registrations
        .map((json) => EventRegistrationModel.fromJson(json as Map<String, dynamic>))
        .toList();
  }

  /// Get registrations by event (admin)
  Future<List<EventRegistrationModel>> fetchByEvent(String eventId) async {
    final response = await _client.get('/event-registrations/event/$eventId');
    final registrations = response['registrations'] as List<dynamic>? ?? [];
    return registrations
        .map((json) => EventRegistrationModel.fromJson(json as Map<String, dynamic>))
        .toList();
  }

  /// Update registration status (admin)
  Future<void> updateStatus(String id, String status) async {
    await _client.patch('/event-registrations/$id/status', {'status': status});
  }

  /// Delete registration (admin)
  Future<void> deleteRegistration(String id) async {
    await _client.delete('/event-registrations/$id');
  }

  /// Get registration count for event
  Future<int> getCount(String eventId) async {
    final response = await _client.get('/event-registrations/$eventId/count');
    return response['count'] as int? ?? 0;
  }
}

class EventRegistrationModel {
  final String id;
  final String eventId;
  final String userId;
  final String status;
  final String? eventTitle;
  final String? eventDate;
  final String? eventLocation;
  final String? eventImageUrl;
  final String? userName;
  final String? userEmail;
  final DateTime? registeredAt;

  EventRegistrationModel({
    required this.id,
    required this.eventId,
    required this.userId,
    required this.status,
    this.eventTitle,
    this.eventDate,
    this.eventLocation,
    this.eventImageUrl,
    this.userName,
    this.userEmail,
    this.registeredAt,
  });

  factory EventRegistrationModel.fromJson(Map<String, dynamic> json) {
    return EventRegistrationModel(
      id: json['id'] as String? ?? '',
      eventId: json['eventId'] as String? ?? '',
      userId: json['userId'] as String? ?? '',
      status: json['status'] as String? ?? 'registered',
      eventTitle: json['eventTitle'] as String?,
      eventDate: json['eventDate'] as String?,
      eventLocation: json['eventLocation'] as String?,
      eventImageUrl: json['eventImageUrl'] as String?,
      userName: json['userName'] as String?,
      userEmail: json['userEmail'] as String?,
      registeredAt: json['registeredAt'] != null 
          ? DateTime.tryParse(json['registeredAt'].toString()) 
          : null,
    );
  }
}
